### ISSUE TITLE
[Feature] Auto completion field

### ISSUE THREAD

**Issue by strarsis** [piklist/piklist#29]
It would be great if an auto completion field (with data source) could be added to piklist fields.

**Comment by JasonTheAdams** [comment]
Hi @strarsis!

I do something very like this using select2 right now. You can pretty easily pull this off right now that way, and can even define your own fields to make it very easily reusable. I'll leave this here for now, but I'm not sure this is something we'll roll into core in the near future as it's currently fairly easily accomplished.

**Comment by manutiedra** [comment]
+1 @strarsis  

Hi @JasonTheAdams, 

do you have any code to show how you integrated in piklist the select2 with ajax to get remote data? 

Even if you don't integrate it in the core (that I think you should) It will help tremendously people like me starting with piklist.

**Comment by JasonTheAdams** [comment]
@manutiedra Sure!

First I enqueue select2 via the [admin_enqueue_scripts](https://codex.wordpress.org/Plugin_API/Action_Reference/admin_enqueue_scripts) hook.

Next, I create my own script which has something like the following:
```js
jQuery(document).ready(function($) {
  $('select.piklist-select2').select2();
}
```

After that, when I'm creating a select field I add the piklist-select2 class to the field:
```php
piklist('field', array(
  'type'      => 'select',
  'field'     => 'select_2_field',
  'label'     => 'My Select2 Field',
  'choices'   => array(
    1           => 'One',
    2           => 'Two'
  ),
  'attributes'=> array(
    'class'     => 'piklist-select2',
    'style'     => 'width: 100%'
  )
));
```

And that's pretty much it! I also have this (which may take some tweaking, and requires piklist to be a dependency of your script when enqueuing) which makes it work with add-more fields, too:
```js
  $('.piklist-field').on('click', '.piklist-field-addmore-wrapper .piklist-addmore-add', function(event) {
    $(event.target).closest('.piklist-field').find('select.piklist-select2').filter(function() {
      return !$(this).data('select2');
    }).select2();
  });
```

Hope this helps! :)

**Comment by manutiedra** [comment]
**Thank you** for the code @JasonTheAdams!

However, I think that the original issue and what I asked you is if there is a way to add select2 with ajax data source (like this: https://select2.org/data-sources/ajax).

For example, if I have a taxonomy called TAX1 with 5000 terms and I want to add a relationship from my CPT to that TAX1, that piklist renders a select2 with autocomplete without loading the 5000 terms, and when the user starts writing, the select2 generates ajax request to get the terms that match what the user was typing. Piklist could be made smart enough to return the data without having to write code and expose a couple of hooks so people can add additional filtering for the data to be returned if needed.

**Comment by JasonTheAdams** [comment]
Hey @manutiedra!

Gotcha. That's certainly an interesting idea, but definitely outside of the scope of Piklist. I'd be interested to look at a PR for something like this. By this point we're talking about working with an advanced feature of another framework entirely that has nothing to do with WordPress. 😄 

With features like this it's advanced enough where someone thinking this far along would probably know how to use Piklist (i.e. create the field with the appropriate class/id), and then add js to hook into the field and instantiate/use Select2. Piklist is already creating the metabox and field, which is create, and the dev can easily go from there. It's cool, but I'm not sure it belongs in core. You could easily create another plugin which extends Piklist (like I did with the [Helper](https://github.com/papertower/piklist-helper)) that makes something like this easy. Maybe a "Piklist Select2" plugin!

**Comment by manutiedra** [comment]
Hi @JasonTheAdams,

I think that the moment piklist offers relations between posts, something like that should be added, because if you have a lot of posts or CPT, relations as they are now because useless. 

Anyway, I'm working in a proof of concept with select2 + REST API for this scenario and I will post the code soon. 

BR, 
Manu

**Comment by JasonTheAdams** [comment]
Interesting point. I can appreciate what you're saying. Personally, since I've only used relate fields admin-side and with post types measuring at most in the 100+ range, I just load (and cache) every post up-front. I really haven't had a use-case otherwise. That said, I can see where much more and AJAX would become a necessity. Still not 100% that it should be in core, as it's still a niche, albeit a good one. An additional plugin may still be the better option.

Curious what @kevinpmiller and @sbruner think?

**Comment by manutiedra** [comment]
Hi @JasonTheAdams,

I've implemented the autocomplete. Please take a look at it:

https://github.com/manutiedra/piklist-autocomplete

and give it a try.

Best regards,
Manu

**Comment by sbruner** [comment]
@manutiedra -- I took a quick look at the field and I like it.

I tested with this code:
`
  piklist('field', array(
    'type' => 'autocomplete',
    'field' => 'autocomplete1',
    'label' => 'autocomplete 1',
    'choices' => array(
        'AL'=>'AL',
        'AK'=>'AK',
        'AZ'=>'AZ',
        'AR'=>'AR',
        'CA'=>'CA',
        'CO'=>'CO',
        'CT'=>'CT',
        'DE'=>'DE',
        'FL'=>'FL',
        'GA'=>'GA',
        'HI'=>'HI',
        'ID'=>'ID',
        'IL'=>'IL',
        'IN'=>'IN',
        'IA'=>'IA',
        'KS'=>'KS',
        'KY'=>'KY',
        'LA'=>'LA',
        'ME'=>'ME',
        'MD'=>'MD',
        'MA'=>'MA',
        'MI'=>'MI',
        'MN'=>'MN',
        'MS'=>'MS',
        'MO'=>'MO',
        'MT'=>'MT',
        'NE'=>'NE',
        'NV'=>'NV',
        'NH'=>'NH',
        'NJ'=>'NJ',
        'NM'=>'NM',
        'NY'=>'NY',
        'NC'=>'NC',
        'ND'=>'ND',
        'OH'=>'OH',
        'OK'=>'OK',
        'OR'=>'OR',
        'PA'=>'PA',
        'RI'=>'RI',
        'SC'=>'SC',
        'SD'=>'SD',
        'TN'=>'TN',
        'TX'=>'TX',
        'UT'=>'UT',
        'VT'=>'VT',
        'VA'=>'VA',
        'WA'=>'WA',
        'WV'=>'WV',
        'WI'=>'WI',
        'WY'=>'WY',
    )
  ));

  piklist('field', array(
    'type' => 'autocomplete',
    'field' => 'autocomplete2',
    'label' => 'autocomplete 2',
    'columns' => 12,
    'options' => array(
      'autocomplete' => array(
        'query' => array(
          'post_type' => 'post'
        ),
      ),
    ),
  ));
`

**Comment by manutiedra** [comment]
@sbruner, I'm glad you like it. I have only tested in the backend, I hope it works in the front end too.

In the second field you posted, the 'autocomplete' data is inside 'options'. It should be one level up. It works because if nothing is specified, the default is to show the posts.

BTW Steve, I've seen in the source that you have extended WP_List_Tables, but I haven't seen any examples of piklist('list_table', ...) in the demos. Do you have an example that you can show?

BR,
Manu

**Comment by JasonTheAdams** [comment]
@manutiedra Out of curiosity, have you tested the field with `'add_more' => true`? Or in a repeater group?

**Comment by manutiedra** [comment]
@JasonTheAdams, yes, I did. However, it doesn't work. I haven't checked the code but it seems that the add more button copies the html but I need to call the select2 init code when that happens. Is there any client side hook for the add more button?

**Comment by kevinpmiller** [comment]
There is indeed.

https://github.com/piklist/piklist/issues/29#issuecomment-358769772

See the code above the selected code for examples of how to handle fields when an add-more is created. 

Let us know if you have any questions!

**Comment by manutiedra** [comment]
Thank you for the info @kevinpmiller 

BTW, even if it is abusing this thread a bit, I have implemented a datatable field:

https://github.com/manutiedra/piklist-datatables

Take a look. 

As it is a read only field (it doesn't post any data), is there any way to avoid including its data in "data-piklist-fields"?

BR,
Manu
